### PR TITLE
move lambda to optional arguments

### PR DESCRIPTION
**Pull request by hongchengyao**

### CODE REVIEW

File: R/ui.R [hunk 1]
```diff
@@ -153,6 +149,15 @@ HarmonyMatrix <- function(
     }
     
     # determine if lambda is appropriate
+    if (!("lambda" %in% names(opt_args))){
+        lambda <- c(0.1, 10)
+    } else{
```
Comment (pati-ni) [inline R/ui.R]: Since we allow only range values of lambda you need to check whether lambda has length 2 before issuing this message.

File: R/ui.R [hunk 2]
```diff
@@ -90,7 +86,7 @@ HarmonyMatrix <- function(
     if (hasArg(do_pca) || hasArg(npcs)) {
         stop('Error: Function parameters do_pca and npcs have been removed in newer versions of harmony. Please remove any of the do_pca or npcs parameters and pass to harmony cell_embeddings directly')
     }
-
+    opt_args <- list(...)
```
Comment (pati-ni) [inline R/ui.R]: It seems that lambda is already set in the `lambda` here if it is passed as a parameter. What is the point of this?

File: R/ui.R [hunk 3]
```diff
@@ -153,6 +149,15 @@ HarmonyMatrix <- function(
     }
     
     # determine if lambda is appropriate
+    if (!("lambda" %in% names(opt_args))){
```
Comment (pati-ni) [inline R/ui.R]: a better function to test this condition is `hasArgs()`

File: R/ui.R [hunk 4]
```diff
@@ -153,6 +149,15 @@ HarmonyMatrix <- function(
...
+        lambda <- c(0.1, 10)
+    } else{
+        message(paste0("How Harmony determine lambda is changed and lambda ",
```
Comment (pati-ni) [inline R/ui.R]: Consider adding that as a warning intead of a message. Also you could wrap the message in a single string so it is consistent with the rest of the package

**Review by pati-ni — CHANGES_REQUESTED** [review]

File: R/ui.R (hunk 2 above)
Comment (hongchengyao) [reply R/ui.R]: I thought `list(...)` is the standard  way to parse ... arguments

File: R/ui.R (hunk 3 above)
Comment (hongchengyao) [reply R/ui.R]: Yes you are right, I will change it. Do you think it's better for me to make all the changes and create a new pull request?

File: R/ui.R (hunk 1 above)
Comment (hongchengyao) [reply R/ui.R]: You mean to do all the lambda integrity checking before the warning message?

File: R/ui.R (hunk 4 above)
Comment (hongchengyao) [reply R/ui.R]: I agree that warning is better, will change this later

File: R/ui.R [hunk 5]
```diff
@@ -75,19 +68,102 @@
...
-                          epsilon.cluster = 1e-5, epsilon.harmony = 1e-4,
-                          plot_convergence = FALSE, return_object = FALSE,
-                          verbose = TRUE, reference_values = NULL, ...) {
+    data_mat, meta_data, vars_use, theta = NULL, sigma = 0.1, nclust = NULL,
+    max_iter = 10, early_stop = TRUE, plot_convergence = FALSE,
+    return_object = FALSE, verbose = TRUE, reference_values = NULL,
+    .options = harmony_options(), ...
+    ) {
 
+    # Parameter checking -------------------------------------------------------
```
Comment (pati-ni) [inline R/ui.R:78]: To minimize the footprint, move the warnings below to a different function and call it like `legacyArgs(...)`

File: R/harmony_option.R [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+
+validate_block.size <- function(block.size){
+    if(block.size <= 0 | block.size > 1){
+        stop('Error: block.size should be set between 0 and 1 (0 < block.size <= 1)')
+    }
+    return(block.size)
+}
+
+validate_max.iter.cluster <- function(max.iter.cluster){
```
Comment (pati-ni) [inline R/harmony_option.R]: Why did you add dummy validate functions?

File: R/ui.R [hunk 7]
```diff
@@ -79,19 +68,102 @@
...
+                   "release of harmony. It will be ignored for this function ",
+                   "call and please remove parameter epsilon.cluster in future ",
+                   "function calls. For advanced users, value of epsilon.cluster ",
+                   "should be adjusted by using parameter .options."),
+            .frequency = "once", .frequency_id = "epsilon.cluster_warning")
+    }
+    if (hasArg(epsilon.harmony)) {
+        rlang::warn(
+            paste0("Warning: Function parameter epsilon.harmony is deprecated in this new ",
+                   "release of harmony. It will be ignored for this function ",
```
Comment (pati-ni) [inline R/ui.R]: Messages could be clearer. Something like the one below is more concise:

Function parameter `` is deprecated, and its user set value will be ignored. Advanced users can set the `` value by passing harmony_options(``) to the .options function parameter.

File: R/harmony_option.R
Comment (pati-ni) [inline R/harmony_option.R:1]: It is not relevant to talk about new releases in the documentation. Remove any mentions of this in the documentation

File: R/harmony_option.R [hunk 8]
```diff
@@ -0,0 +1,98 @@
+#' Set advanced options for HarmonyMatrix
+#' @param lambda Default lambda = c(0.1, 10). In the new release of Harmony,
```
Comment (pati-ni) [inline R/harmony_option.R]: The docstring of lambda is not clear.

1. Describe what lambda corresponds to.
2. Mention that this parameter accepts a range and the parameter is being set dynamically using heuristics during runtime.
3. Mention that a trivial range [1,1] will disable the heuristics and lambda will be fixed.

File: R/ui.R [hunk 9]
```diff
@@ -47,6 +33,9 @@
 #' @param reference_values (Advanced Usage) Defines reference
 #'     dataset(s).  Cells that have batch variables values matching
 #'     reference_values will not be moved.
+#' @param .options Advanced parameters of HarmonyMatrix. This must be the
```
Comment (pati-ni) [inline R/ui.R]: Add a small example on some valid form of the `.options` parameter so users do not have to look it up separately.

**Review by pati-ni — CHANGES_REQUESTED** [review]

File: R/ui.R (hunk 5 above)
Comment (hongchengyao) [reply R/ui.R:78]: Good idea, will do it

File: R/harmony_option.R (hunk 6 above)
Comment (hongchengyao) [reply R/harmony_option.R]: Ilya thought we may want to validate every parameter in the future, so I just add them for future extension

**Comment by pati-ni** [comment]
Sure, add it when we are going to add it. The infrastructure is there
already. No reason to have code that does nothing

File: R/ui.R (hunk 7 above)
Comment (hongchengyao) [reply R/ui.R]: Sounds good

File: R/harmony_option.R
Comment (hongchengyao) [reply R/harmony_option.R:1]: Sure

File: R/harmony_option.R (hunk 8 above)
Comment (hongchengyao) [reply R/harmony_option.R]: OK

File: R/ui.R (hunk 9 above)
Comment (hongchengyao) [reply R/ui.R]: I'm not sure about this. In general, we don't want users to touch parameters set by .options. And if they really want to do that as advanced users, they should really refer to the docs of harmony_options()

File: R/ui.R (hunk 5 above)
Comment (pati-ni) [reply R/ui.R:78]: I meant to copy-paste the previous block in a function where you pass all the unknown parameters `...`. This also works ok I guess it is 8 lines of code.

File: R/ui.R [hunk 10]
```diff
@@ -216,3 +206,52 @@ HarmonyMatrix <- function(
     }
 }
 
+
+legacy_args <- function(param){
```
Comment (pati-ni) [inline R/ui.R:210]: Move this to harmony_options.R?
